[Merged by Bors] - feat(Analysis/SpecialFunctions): add the sigmoid function

gaetanserre opened the pull request:
Add the Sigmoid from `ℝ` to `ℝ` and from `ℝ` to `I` and show some analytical properties. Then, add `sigmoid_ord_embedding`, the Sigmoid function seen as an `OrderEmbedding` from `ℝ` to `I`. Show that it is a `MeasurableEmbedding` and that we could use it to construct a `MeasurableEmbedding` from any `StandardBorelSpace` to `I`.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

gaetanserre committed:
feat(Analysis/SpecialFunctions): Add the Sigmoid function as an `OrderEmbedding`

Add the Sigmoid function as an `OrderEmbedding` from `ℝ` to `I`. Show
that Sigmoid is a `MeasurableEmbedding` and that we could use it to
construct a `MeasurableEmbedding` from any `StandardBorelSpace` to `I`.

RemyDegenne committed:
add Rémy Degenne as coauthor

gaetanserre commented:
I created a new file for Sigmoid (Analysis/SpecialFunctions/Sigmoid) but it might not be the best place to put these results.

gaetanserre committed:
Add docstring for the Sigmoid definition

Define the Sigmoid function as an OrderEmbedding from ℝ to I.

themathqueen commented on Mathlib/Analysis/SpecialFunctions/Sigmoid.lean:
```diff
@@ -0,0 +1,93 @@
+  rw [range_eq]
+  exact Ioo_subset_Ioo a.2.1 b.2.2
+
+lemma measurable_real_embedding : MeasurableEmbedding Sigmoid := by
+  refine is_embedding.measurableEmbedding ?_
+  rw [range_eq]
+  exact measurableSet_Ioo
```
```suggestion
lemma measurable_real_embedding : MeasurableEmbedding Sigmoid :=
  is_embedding.measurableEmbedding <| range_eq ▸ measurableSet_Ioo
```

themathqueen commented on Mathlib/Analysis/SpecialFunctions/Sigmoid.lean:
```diff
@@ -0,0 +1,93 @@
+    suffices exp (log (-1 + x.1⁻¹)) = -1 + x.1⁻¹ by
+      rw [this]
+      simp only [add_neg_cancel_left, inv_inv]
+    refine exp_log (lt_neg_add_iff_lt.mpr ?_)
+    exact one_lt_inv_iff₀.mpr hx
```
```suggestion
    exact exp_log (lt_neg_add_iff_lt.mpr <| one_lt_inv_iff₀.mpr hx)
```

themathqueen commented on Mathlib/Analysis/SpecialFunctions/Sigmoid.lean:
```diff
@@ -0,0 +1,93 @@
+    refine exp_log (lt_neg_add_iff_lt.mpr ?_)
+    exact one_lt_inv_iff₀.mpr hx
+
+lemma is_embedding : Topology.IsEmbedding Sigmoid := by
+  refine Sigmoid.isEmbedding_of_ordConnected (ordConnected_of_Ioo ?_)
+  intro a ha b hb hab
+  rw [range_eq]
+  exact Ioo_subset_Ioo a.2.1 b.2.2
```
```suggestion
lemma is_embedding : Topology.IsEmbedding Sigmoid :=
  Sigmoid.isEmbedding_of_ordConnected (ordConnected_of_Ioo <|
    fun a _ b _ _ => range_eq ▸ Ioo_subset_Ioo a.2.1 b.2.2)
```

themathqueen commented on Mathlib/Analysis/SpecialFunctions/Sigmoid.lean:
```diff
@@ -0,0 +1,93 @@
+    apply Subtype.eq
+    simp only
+    suffices exp (log (-1 + x.1⁻¹)) = -1 + x.1⁻¹ by
+      rw [this]
+      simp only [add_neg_cancel_left, inv_inv]
```
```suggestion
      simp only [this, add_neg_cancel_left, inv_inv]
```

themathqueen commented on Mathlib/Analysis/SpecialFunctions/Sigmoid.lean:
```diff
@@ -0,0 +1,93 @@
+    use -(log (-1 + x.1⁻¹))
+    simp only [Sigmoid, RelEmbedding.coe_mk, Embedding.coeFn_mk, neg_neg]
+    apply Subtype.eq
+    simp only
```
```suggestion
```

themathqueen commented on Mathlib/Analysis/SpecialFunctions/Sigmoid.lean:
```diff
@@ -0,0 +1,93 @@
+    rw [mem_range]
+    use -(log (-1 + x.1⁻¹))
+    simp only [Sigmoid, RelEmbedding.coe_mk, Embedding.coeFn_mk, neg_neg]
+    apply Subtype.eq
```
```suggestion
```

themathqueen commented on Mathlib/Analysis/SpecialFunctions/Sigmoid.lean:
```diff
@@ -0,0 +1,93 @@
+    refine ⟨(1 + exp (-a))⁻¹, ?_⟩
+    constructor
+    · positivity
+    · refine inv_le_one_of_one_le₀ ?_
+      exact (le_add_iff_nonneg_right 1).mpr (exp_nonneg (-a))
```
```suggestion
    · exact inv_le_one_of_one_le₀ <| (le_add_iff_nonneg_right 1).mpr (exp_nonneg (-a))
```

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: themathqueen <23701951+themathqueen@users.noreply.github.com>

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: themathqueen <23701951+themathqueen@users.noreply.github.com>

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: themathqueen <23701951+themathqueen@users.noreply.github.com>

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: themathqueen <23701951+themathqueen@users.noreply.github.com>

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: themathqueen <23701951+themathqueen@users.noreply.github.com>

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: themathqueen <23701951+themathqueen@users.noreply.github.com>

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: themathqueen <23701951+themathqueen@users.noreply.github.com>

gaetanserre commented:
@themathqueen thank you for your suggestions!

themathqueen commented on Mathlib/Analysis/SpecialFunctions/Sigmoid.lean:
```diff
@@ -0,0 +1,84 @@
+
+/-- The Sigmoid function as an `OrderEmbedding` from `ℝ` to `I`. -/
+noncomputable def Sigmoid : ℝ ↪o I where
+  toFun a := by
+    refine ⟨(1 + exp (-a))⁻¹, ?_⟩
+    constructor
+    · positivity
+    · exact inv_le_one_of_one_le₀ <| (le_add_iff_nonneg_right 1).mpr (exp_nonneg (-a))
```
```suggestion
  toFun a := ⟨(1 + exp (-a))⁻¹, by positivity,
    inv_le_one_of_one_le₀ <| (le_add_iff_nonneg_right 1).mpr (exp_nonneg (-a))⟩
```

themathqueen commented on Mathlib/Analysis/SpecialFunctions/Sigmoid.lean:
```diff
@@ -0,0 +1,84 @@
+    simp_all only [Subtype.mk.injEq, inv_inj, add_right_inj, exp_eq_exp, neg_inj]
+  map_rel_iff' {a b} := by
+    simp_all only [Embedding.coeFn_mk, Subtype.mk_le_mk]
+    refine ⟨fun h ↦ ?_, fun h ↦ ?_⟩
+    · suffices -b ≤ -a from neg_le_neg_iff.mp this
+      rwa [←exp_le_exp, ←add_le_add_iff_left 1, ←inv_le_inv₀ (by positivity) (by positivity)]
+    · gcongr
```
to make it more compact:
```suggestion
    refine ⟨fun h ↦ ?_, fun h ↦ by gcongr⟩
    suffices -b ≤ -a from neg_le_neg_iff.mp this
    rwa [←exp_le_exp, ←add_le_add_iff_left 1, ←inv_le_inv₀ (by positivity) (by positivity)]
```

j-loreaux reviewed (commented):
So, I think it might make more sense to first develop some theory for the logistic function(s) (as a function from `ℝ → ℝ`), and then specialize that to prove some of these properties of the sigmoid. For example, you could establish its asymptotic behavior.

themathqueen commented on Mathlib/Analysis/SpecialFunctions/Sigmoid.lean:
```diff
@@ -0,0 +1,84 @@
+lemma range_eq : range Sigmoid = Ioo 0 1 := by
+  ext x
+  constructor
+  · intro hx
+    obtain ⟨y, rfl⟩ := hx
```
```suggestion
  · rintro ⟨y, rfl⟩
```

themathqueen commented on Mathlib/Analysis/SpecialFunctions/Sigmoid.lean:
```diff
@@ -0,0 +1,84 @@
+  · intro hx
+    obtain ⟨y, rfl⟩ := hx
+    constructor
+    · simp only [Sigmoid, RelEmbedding.coe_mk, Embedding.coeFn_mk]
+      change 0 < (1 + exp (-y))⁻¹
```
```suggestion
    · change 0 < (1 + exp (-y))⁻¹
```

themathqueen commented on Mathlib/Analysis/SpecialFunctions/Sigmoid.lean:
```diff
@@ -0,0 +1,84 @@
+    · simp only [Sigmoid, RelEmbedding.coe_mk, Embedding.coeFn_mk]
+      change 0 < (1 + exp (-y))⁻¹
+      positivity
+    · simp only [Sigmoid]
+      change (1 + exp (-y))⁻¹ < 1
+      refine inv_lt_one_of_one_lt₀ ?_
+      exact lt_add_of_pos_right 1 (exp_pos (-y))
```
```suggestion
    · exact inv_lt_one_of_one_lt₀ <| lt_add_of_pos_right 1 (exp_pos (-y))
```

themathqueen commented on Mathlib/Analysis/SpecialFunctions/Sigmoid.lean:
```diff
@@ -0,0 +1,84 @@
+      refine inv_lt_one_of_one_lt₀ ?_
+      exact lt_add_of_pos_right 1 (exp_pos (-y))
+  · intro hx
+    rw [mem_range]
```
```suggestion
```

themathqueen commented on Mathlib/Analysis/SpecialFunctions/Sigmoid.lean:
```diff
@@ -0,0 +1,84 @@
+  · intro hx
+    rw [mem_range]
+    use -(log (-1 + x.1⁻¹))
+    simp only [Sigmoid, RelEmbedding.coe_mk, Embedding.coeFn_mk, neg_neg]
```
```suggestion
```

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: Jireh Loreaux <loreaujy@gmail.com>

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: Jireh Loreaux <loreaujy@gmail.com>

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: Jireh Loreaux <loreaujy@gmail.com>

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: Jireh Loreaux <loreaujy@gmail.com>

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: Jireh Loreaux <loreaujy@gmail.com>

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: themathqueen <23701951+themathqueen@users.noreply.github.com>

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: themathqueen <23701951+themathqueen@users.noreply.github.com>

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: themathqueen <23701951+themathqueen@users.noreply.github.com>

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: themathqueen <23701951+themathqueen@users.noreply.github.com>

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: themathqueen <23701951+themathqueen@users.noreply.github.com>

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: themathqueen <23701951+themathqueen@users.noreply.github.com>

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: themathqueen <23701951+themathqueen@users.noreply.github.com>

gaetanserre committed:
Add analytical properties of sigmoid

j-loreaux reviewed (commented):
Thanks for expanding the API here. It's certainly progressing.

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: Jireh Loreaux <loreaujy@gmail.com>

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: Jireh Loreaux <loreaujy@gmail.com>

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: Jireh Loreaux <loreaujy@gmail.com>

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: Jireh Loreaux <loreaujy@gmail.com>

gaetanserre committed:
Add TODO

gaetanserre committed:
Add `sigmoid` as a function from `ℝ` to `I`

gaetanserre committed:
docstring

gaetanserre committed:
ring_nf

gaetanserre committed:
sigmoid is analytic and contdiff

gaetanserre commented on Mathlib/Analysis/SpecialFunctions/Sigmoid.lean:
```diff
@@ -0,0 +1,197 @@
+
+@[simp]
+theorem differentiable_sigmoid : Differentiable 𝕜 sigmoid := fun x =>
+  (hasDerivAt_sigmoid x).differentiableAt.restrictScalars 𝕜
```
I just added those results ([#c5649ab](https://github.com/leanprover-community/mathlib4/pull/28780/commits/c5649abdad96d58d82675b1b96761d1e2663e72d)).

gaetanserre commented:
Thank you for all your comments and suggestions, that is my first time creating a new Mathlib file.

> You shouldn't be developing the unitInterval version before the Real version. You should be building up structure as you go. You'll still be able to prove the unitInterval versions from the Real versions trivially in most cases.

I am not sure to understand this, though. What I understand is that I should first define `Real.sigmoid` then `unitInterval.sigmoid` using `Subtype.coind` and then adding lemmas about these two functions without splitting the ones concerning the real version of sigmoid from the unitInterval one, but instead grouping the lemmas that have the same statement (one for the real version and one for unitInterval version). Is that correct?

j-loreaux commented:
I just mean you should define `Real.sigmoid` first and price all your results about it. Then you define `unitInterval.sigmoid` and prove some results about that.

Currently you have done it backwards. Does that make sense?

gaetanserre commented:
Yeah it makes sense, thanks a lot. I'll do that asap.

gaetanserre committed:
Merge branch 'leanprover-community:master' into Sigmoid

gaetanserre committed:
Reorganize

gaetanserre commented:
I tried to do what you suggested. The only thing is that I don't know where I should add the lemma `Subtype.range_coind`, so for now I kept the old proof. Do you have any idea of where I should put this lemma? I can't directly add it in `Data/Subtype.lean` as some imports are missing and I don't want to mess with that.

j-loreaux reviewed (commented):
What does `#find_home` say for `Subtype.range_coind`? My bet is that it can go in `Data.Set.Operations`.

j-loreaux commented:
Also, please mark conversations as "resolved" on GitHub if you think you have addressed them.

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: Jireh Loreaux <loreaujy@gmail.com>

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: Jireh Loreaux <loreaujy@gmail.com>

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: Jireh Loreaux <loreaujy@gmail.com>

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: Jireh Loreaux <loreaujy@gmail.com>

gaetanserre committed:
Subtype.range_coind

gaetanserre commented:
> What does `#find_home` say for `Subtype.range_coind`? My bet is that it can go in `Data.Set.Operations`.

Indeed, `#find_home` returned `Data.Set.Operations`, so I added it here and modified the proof accordingly.

gaetanserre committed:
CI

gaetanserre committed:
Merge branch 'leanprover-community:master' into Sigmoid

j-loreaux reviewed (commented):
Thanks! I think we're really close here.

gaetanserre committed:
Update Mathlib/Data/Set/Operations.lean

Co-authored-by: Jireh Loreaux <loreaujy@gmail.com>

gaetanserre committed:
Update Mathlib/Analysis/SpecialFunctions/Sigmoid.lean

Co-authored-by: Jireh Loreaux <loreaujy@gmail.com>

gaetanserre committed:
Add composition theorems

RemyDegenne reviewed (approved):
Thanks!
bors r+
